### PR TITLE
rsx: Restructure ZCULL handling

### PR DESCRIPTION
**Pull request by kd-11**
The motivation behind this one is interesting. The reason ZCULL emulation is so slow is that NVIDIA lumped in timestamp reading with passing occlusion samples as one object called a report. So if a game is just using the reports to tell the RSX time or check on RSX progress using timestamps we cannot tell if it wants to read the timestamp or wants to know how many samples have been rendered so far. Timestamps are very cheap to calculate, but counting samples is very slow and punishing as we have to send work to the GPU and wait for the results to come back. However, while making a hacked build for debugging, I realized something - most games use a library call to read timestamps despite the fact that they can just read the values themselves! This gives us a semi-reliable way of knowing if the game just wanted to see the current timestamp allowing us to skip doing the expensive sample evaluation every time.

Changes:
- Force cellGcmGetTimestamp to be HLE. This allows us to use super memory to read timestamp values without disturbing ZCULL behavior. Of course some games will not use this method, so we need page locking.
- Lock active ZCULL memory. In many games, we're not even seeing faults in this region, because they aren't actually doing anything with the report data. They just wanted the timestamps, which we can check for cheaply using cellGcmGetTimestamp.
- If a game requests for conditional rendering, try to calculate it on the GPU as much as possible. It is faster to just assume it is visible and let the GPU do the culling work instead of trying to emulate it on the CPU.
- Work analysis in several titles showed that calling ZCULL_control::update every RSX cycle is a very poor way of doing work efficiently. Doing it every 64 loop cycles is a lot more efficient and the duration between calls is actually large enough that we get a much better hit ratio. This improves CPU efficiency by a lot.
- Minor optimization: Get rid of std::multimap for image view tables. It is easier to just encode for what we want.

### CODE REVIEW

File: rpcs3/Emu/RSX/RSXZCULL.cpp [hunk 1]
```diff
@@ -0,0 +1,992 @@
...
+
+		void ZCULL_control::set_status(class ::rsx::thread* ptimer, bool surface_active, bool zpass_active, bool zcull_stats_active, bool flush_queue)
+		{
+			write_enabled = surface_active;
+			zpass_count_enabled = zpass_active;
+			stats_enabled = zcull_stats_active;
+
+			check_state(ptimer, flush_queue);
+
+			// Disabled since only ZPASS is implemented right now
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/RSXZCULL.cpp:108]: maybe add a TODO here

**Comment by cipherxof** [comment]
50% increase in MGS4.

**Master:**
master

**PR:**
pr

File: rpcs3/Emu/RSX/RSXZCULL.cpp [hunk 2]
```diff
@@ -0,0 +1,992 @@
...
+		}
+
+		void conditional_render_eval::eval_result(::rsx::thread* pthr)
+		{
+			vm::ptr<CellGcmReportData> result = vm::cast(eval_address);
+			const bool failed = (result->value == 0u);
+			set_eval_result(pthr, failed);
+		}
+	}
+}
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/RSXZCULL.cpp:992]: mssing newline

**Comment by Blackbird88** [comment]
MGS4 crashes with this error
`F {RSX [0x0a11438]} SIG: Thread terminated due to fatal error: Verification failed
(in file C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\rpcs3\Emu\RSX\RSXThread.cpp:1752[:3], in function get_current_vertex_program)`

**Comment by kd-11** [comment]
> [RPCS3.zip](https://github.com/RPCS3/rpcs3/files/8702433/RPCS3.zip) Killzone 2 crashes with verification failed error few seconds after going ingame

Confirmed, I'm investigating.

**Comment by Joaozin-tech** [comment]
My Test INFAMOUS 1 Speed I5 11400F Thanks !!!!! @kd-11

**Comment by kd-11** [comment]
R&C (and infamous to some extent) have other issues related to high draw call count. I'm aware of the problems and investigating solutions.

**Comment by Jonathan44062** [comment]
Sonic Unleashed crashes with 
Verification failed
(in file C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\rpcs3\Emu\RSX\RSXThread.cpp:1954[:3], in function get_current_fragment_program)
A few minutes of gameplay 
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/8702843/RPCS3.log)

**Comment by Dormant-Hero** [comment]
Okay ignore my uncharted 2 comment. Seems to work with the latest version of the PR just posted for windows as per the below:

image

**Comment by Xcedf** [comment]
Got this error in GTA V
F {RSX [0x0a11438]} SIG: Thread terminated due to fatal error: Verification failed (in file C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\rpcs3\Emu\RSX\RSXThread.cpp:1752[:3], in function get_current_vertex_program)

**Comment by kd-11** [comment]
I know what's causing it, I'm working on fixing it without hurting performance.

**Comment by kd-11** [comment]
A tentative fix is up for the crash. Let me know if you're still getting crashing once its done building.

**Comment by Darkhost1999** [comment]
This appears to have fixed the crash I was experiencing when testing


For the curious "does this fix GTA V disable zcull occlusion queries"
I tested on the first commit and it looked like

So I suspect it's a different issue.

**Comment by Xcedf** [comment]
@kd-11 tested for half an hour in gta v crash seems gone
@Darkhost1999 it is a different issue, disabling ZCull just removes the water reflection layer, which has this issue

**Comment by CatKasha** [comment]
comparison with v0.0.22-13600-2ba437b6 and v0.0.22-477dc0f5
clean install, 4.89, default settings + performance overlay, 3700x stock, gtx760 (472.98)

DiRT™ Demo [NPUB90042]
same fps
loading time has increased
on both version, crash after force close when stuck on loading screen (static image), to reproduce it play first event, then open latest (ill create issue with logs later)

DiRT 2 Demo [NPUB90294] - no regression and crashes, same fps with unlocked fps

METAL GEAR SOLID 4 DEMO VERSION [NPUB90176] - tested only first area when player got control, cant tell about stability
fps increase with with unlocked fps
v0.0.22-13600-2ba437b6
main
pr v0.0.22-477dc0f5
pr

v0.0.22-13600-2ba437b6
main2
pr v0.0.22-477dc0f5
pr2

**Comment by Linear524** [comment]
@cipherxof, how did you fix MGS4 z-fighting issue on master builds ? Far ground textures is always flickering for me...
I heard that it is an attribute precision issues with Z-fighting problem.

**Comment by kd-11** [comment]
> @cipherxof, how did you fix MGS4 z-fighting issue on master builds ? Far ground textures is always flickering for me... I heard that it is an attribute precision issues with Z-fighting problem.

They don't. Custom builds use an old hack with exaggerated polygon offset.

**Comment by Linear524** [comment]
All latest version keeps freezing on GoW3 and Ratchet & Clank titles...
No log errors, just sudden freeze like RSX dies silently :(
Other games runs fine (Killzone, Gran Turismo 6, Persona5, TLoU, etc...

**Comment by kd-11** [comment]
> All latest version keeps freezing on GoW3 and Ratchet & Clank titles... No log errors, just sudden freeze like RSX dies silently :( Other games runs fine (Killzone, Gran Turismo 6, Persona5, TLoU, etc...

Please open a proper bug report ticket with appropriate logs and steps to reproduce.

**Comment by tamburro92** [comment]
Introduced a new bug on mac os:
#12158 https://github.com/RPCS3/rpcs3/issues/12158

**Pull request merged**